src/reducers/facetconfig.ts

```diff
@@ -0,0 +1,148 @@
+import { FacetConfig } from '../typings/schema';
+import { NormalizedFacetConfig } from '../typings/index';
+import { arrayMove } from 'react-sortable-hoc';
+
+// state def
+export type FacetConfigReducer = NormalizedFacetConfig[];
+const defaultState: FacetConfigReducer = [];
+
+// actions
+type AddFacetConfigItemAction = {
+    type: 'ADD_FACET_CONFIG_ITEM',
+    payload: {
+        facetConfig: FacetConfig
+    }
+};
+type SetFacetConfigItemsAction = {
+    type: 'SET_FACET_CONFIG_ITEMS',
+    payload: {
+        facetConfigs: FacetConfig[]
+    }
+};
+
+type DeleteFacetConfigItemAction = {
+    type: 'DELETE_FACET_CONFIG_ITEM',
+    payload: {
+        id: number
+    }
+};
+
+type ModifyFacetConfigItemAction = {
+    type: 'MODIFY_FACET_CONFIG_ITEM',
+    payload: {
+        id: number,
+        facetConfig: FacetConfig
+    }
+};
+
+type SortFacetConfigItemAction = {
+    type: 'SORT_FACET_CONFIG_ITEM',
+    payload: {
+        oldIndex: number,
+        newIndex: number
+    }
+};
+
+type Action =
+    AddFacetConfigItemAction
+    | SetFacetConfigItemsAction
+    | DeleteFacetConfigItemAction
+    | ModifyFacetConfigItemAction
+    | SortFacetConfigItemAction;
+
+// selectors
+
+export const lastId = (state: FacetConfigReducer): number => state
+    .map(config => config.id)
+    .reduce(((previousValue: number, currentValue: number) => Math.max(previousValue, currentValue)), -1);
+
+export const getById = (id: number, state: FacetConfigReducer): NormalizedFacetConfig | undefined =>
+    state.find(config => config.id === id);
+
+// reducer
+
+const item = (state: NormalizedFacetConfig | null, action: Action, items: NormalizedFacetConfig[]): NormalizedFacetConfig => {
+    switch (action.type) {
+        case 'ADD_FACET_CONFIG_ITEM':
+            return {
+                id: lastId(items) + 1,
+                ...action.payload.facetConfig
+            };
+        case 'MODIFY_FACET_CONFIG_ITEM':
+            return {
+                id: state!.id,
+                ...action.payload.facetConfig
```

Comment:
Spread operators should always be above the new value you're providing, just to make sure it gets overwritten by the new value your pushing right?